### PR TITLE
Commander: rework GPS quality thresholds for Home position setting and GPS invalid warning

### PR DESCRIPTION
**Pull request by sfuhrer**
### Solved Problem
- checks for GPS validity in Commander (used for setting Home and for triggering warning if GPS data is bad) are linked to failsafe thresholds (eg. COM_POS_FS_EPH), so don't reflect if the GPS is currently used or not by the estimator

### Solution
- decouple GPS checks for Home position setting (with hard coded thresholds)
- change "GPS invalid" warning to "GPS stopped being fused warning", that acts on real estimator feedback not separate thresholds

Previously failsafe_flags.gps_position_invalid was used for both. Beside it being two different things that really shouldn't have the same thresholds, it for me also doesn't belong into the failsafe flags as it doesn't trigger any failsafe (only losing global/local position does). 

One diff in the Home position setting is that I dropped (for now) the valid hysteresis around _vehicle_gps_position_valid, which required the checks to pass for 3s before Home position was set. Do we want that? 

### Alternatives
Certainly can be architecture differently, please advice. 

### Test coverage
SITL tested:
Triggered gps failure through failure injection, made it ok again, made it fail again, and then wait till (leave vehicle dead-reckoning) till position estimate inaccuracy grows above COM_POS_LOW_EPH and system RTLs.
https://user-images.githubusercontent.com/26798987/220408751-b46aafd2-efdb-44d6-9932-050a8592a6e3.mp4
image

### Context
This came up as a follow up for https://github.com/PX4/PX4-Autopilot/pull/21018.

### CODE REVIEW

File: src/modules/commander/HealthAndArmingChecks/checks/estimatorCheck.cpp [hunk 1]
```diff
@@ -280,15 +273,38 @@ void EstimatorChecks::checkEstimatorStatus(const Context &context, Report &repor
...
+					     "GNSS data fusion stopped");
+
+			} else if (!_gps_was_fused && ekf_gps_fusion) {
+
+				if (reporter.mavlink_log_pub()) {
+					mavlink_log_info(reporter.mavlink_log_pub(), "GNSS data fusion started\t");
+				}
+
+				events::send(events::ID("check_estimator_gnss_fusion_stated"), {events::Log::Info, events::LogInternal::Info},
+					     "GNSS data fusion stated");
```
Comment (bkueng) [inline src/modules/commander/HealthAndArmingChecks/checks/estimatorCheck.cpp:301]: `stated` -> `started`

**Review by bkueng — APPROVED** [review]

File: src/modules/commander/HealthAndArmingChecks/checks/estimatorCheck.cpp [hunk 2]
```diff
@@ -798,24 +799,6 @@ void EstimatorChecks::setModeRequirementFlags(const Context &context, bool pre_f
...
 	failsafe_flags.angular_velocity_invalid = angular_velocity_invalid;
-
-
-	// gps
-	if (vehicle_gps_position.timestamp != 0) {
-		bool time = (now < vehicle_gps_position.timestamp + 1_s);
-
-		bool fix = vehicle_gps_position.fix_type >= 2;
-		bool eph = vehicle_gps_position.eph < _param_com_pos_fs_eph.get();
-		bool epv = vehicle_gps_position.epv < _param_com_pos_fs_epv.get();
```
Comment (bresch) [inline src/modules/commander/HealthAndArmingChecks/checks/estimatorCheck.cpp:809]: This was the only place where `com_pos_fs_epv` was used, we can get rid of the parameter

**Review by bresch — CHANGES_REQUESTED** [review]

File: src/modules/commander/HealthAndArmingChecks/checks/estimatorCheck.cpp (hunk 2 above)
Comment (sfuhrer) [reply src/modules/commander/HealthAndArmingChecks/checks/estimatorCheck.cpp:809]: Good call, addressed in 3e5f23c0fddaeb1dac75aa3060cc417c06490593

File: src/modules/commander/HomePosition.cpp [hunk 3]
```diff
@@ -314,6 +316,19 @@ void HomePosition::update(bool set_automatically, bool check_if_changed)
 	_local_position_sub.update();
 	_global_position_sub.update();
 
+	if (_vehicle_gps_position_sub.updated()) {
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/commander/HomePosition.cpp:303]: Do you prefer 2dd20134ead924ca5de9667a40f97969446ddaae?

[the remark this answers is not in the document]
Comment (bkueng) [reply src/modules/commander/HomePosition.cpp:303]: Yes that works

**Review by bkueng — DISMISSED** [review]

File: src/modules/commander/HomePosition.hpp [hunk 4]
```diff
@@ -75,4 +80,10 @@ class HomePosition
 	uint8_t							_heading_reset_counter{0};
 	bool							_valid{false};
 	const failsafe_flags_s					&_failsafe_flags;
+	bool							_gps_position_for_home_valid{false};
+	double							_gps_lat{0};
```
Comment (bresch) [inline src/modules/commander/HomePosition.hpp:84]: We could almost change the GPS sub to `SubscriptionData`

**Review by bresch — APPROVED** [review]

File: src/modules/commander/HomePosition.hpp (hunk 4 above)
Comment (sfuhrer) [reply src/modules/commander/HomePosition.hpp]: Like 72260d00866d0aa39aa20ee434e947b13bfc2201?

File: src/modules/commander/HomePosition.cpp [hunk 5]
```diff
@@ -313,6 +305,18 @@ void HomePosition::update(bool set_automatically, bool check_if_changed)
 {
 	_local_position_sub.update();
 	_global_position_sub.update();
+	_vehicle_gps_position_sub.update();
+
+	if (_vehicle_gps_position_sub.updated()) {
```
Comment (sfuhrer) [inline src/modules/commander/HomePosition.cpp:310]: @bkueng is `.updated() `on `SubscriptionData` working the same as on normal Subscriptions?

Comment (bkueng) [reply src/modules/commander/HomePosition.cpp:310]: Yes but it's better to store only the few fields that you need.

File: src/modules/commander/HomePosition.hpp (hunk 4 above)
Comment (sfuhrer) [reply src/modules/commander/HomePosition.hpp:84]: Reverted https://github.com/PX4/PX4-Autopilot/commit/72260d00866d0aa39aa20ee434e947b13bfc2201 again.

File: src/modules/commander/HomePosition.cpp (hunk 5 above)
Comment (sfuhrer) [reply src/modules/commander/HomePosition.cpp:303]: reverted https://github.com/PX4/PX4-Autopilot/commit/72260d00866d0aa39aa20ee434e947b13bfc2201

**Comment by dakejahl** [comment]
@sfuhrer I was quite surprised to discover that the home position was being set before the estimator GPS checks pass. This results in a degraded UX -- I hear the tune indicating the GPS home position is set, but then I still get arming denied for a few seconds until the estimator checks pass.

I think setting the home position before GPS checks pass isn't necessarily wrong, so perhaps we should just change the purpose of the tune? It is useful to know when GPS is ready to fly, I think the tune is better suited for this. What do you think?

**Comment by sfuhrer** [comment]
> It is useful to know when GPS is ready to fly, I think the tune is better suited for this. What do you think?

Generally agree, but not strongly opinionated. "ready to fly in a position-controlled mode" specifically. 
I guess the counter argument is that some people just care about Home being set or not such that they have a safe RTL point after taking off in a non-position-controlled mode.

**Pull request merged**